File: perl/example_authenticate.pl
#!/usr/bin/perl

use strict;
use Tinkerforge::IPConnection;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant SECRET => 'My Authentication Secret!';

# Create IPConnection
our $ipcon = Tinkerforge::IPConnection->new();

# Disable auto reconnect mechanism, in case we have the wrong secret.
# If the authentication is successful, reenable it.
$ipcon->set_auto_reconnect(0);

# Authenticate each time the connection got (re-)established
sub cb_connected
{
	my ($connect_reason) = @_;

	if ($connect_reason == $ipcon->CONNECT_REASON_REQUEST)
	{
		print "Connected by request\n";
	}
	elsif ($connect_reason == $ipcon->CONNECT_REASON_AUTO_RECONNECT)
	{
		print "Auto-Reconnect\n";
	}

	# Authenticate first...
	eval
	{
		$ipcon->authenticate(&SECRET);
		print "Authentication succeeded\n";
	};
	if ($!)
	{
		print "Could not authenticate: $!\n";
		return;
	}

	# ...reenable auto reconnect mechanism, as described above...
	$ipcon->set_auto_reconnect(1);

	# ...then trigger Enumerate
	$ipcon->enumerate();
}

# Print incoming enumeration
sub cb_enumerate
{
	my ($uid, $connected_uid, $position, $hardware_version,
	    $firmware_version, $device_identifier, $enumeration_type) = @_;

	print "UID: $uid, Enumeration Type: $enumeration_type\n";
}

# Register Connected Callback
$ipcon->register_callback($ipcon->CALLBACK_CONNECTED, 'cb_connected');

# Register Enumerate Callback
$ipcon->register_callback($ipcon->CALLBACK_ENUMERATE, 'cb_enumerate');

# Connect to brickd
$ipcon->connect(&HOST, &PORT);

print "Press any key to exit...\n";
<STDIN>;
$ipcon->disconnect();

File: javascript/ExampleAuthenticate.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var SECRET = 'My Authentication Secret!';

ipcon = new Tinkerforge.IPConnection(); // Create IP connection
ipcon.connect(HOST, PORT,
    function(error) {
        console.log('Error: '+error);
    }
); // Connect to brickd

// Disable auto reconnect mechanism, in case we have the wrong secret.
// If the authentication is successful, reenable it.
ipcon.setAutoReconnect(false);

// Register Connected Callback
ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    // Authenticate each time the connection got (re-)established
    function(connectReason) {
        switch(connectReason) {
        case Tinkerforge.IPConnection.CONNECT_REASON_REQUEST:
            console.log('Connected by request');
            break;
        case Tinkerforge.IPConnection.CONNECT_REASON_AUTO_RECONNECT:
            console.log('Auto-Reconnected');
            break;
        }
        ipcon.authenticate(SECRET,
            function() {
                console.log('Authentication succeeded');

                // ...reenable auto reconnect mechanism, as described above...
                ipcon.setAutoReconnect(true);

                // ...then trigger Enumerate
                ipcon.enumerate();
            },
            function(error) {
                console.log('Could not authenticate: '+error);
            }
        );
    }
);

// Register Enumerate Callback
ipcon.on(Tinkerforge.IPConnection.CALLBACK_ENUMERATE,
    // Print incoming enumeration
    function(uid, connectedUid, position, hardwareVersion, firmwareVersion,
             deviceIdentifier, enumerationType) {
        console.log('UID: '+uid+', Enumeration Type: '+enumerationType);
    }
);

console.log("Press any key to exit ...");
process.stdin.on('data',
    function(data) {
        ipcon.disconnect();
        process.exit(0);
    }
);
